### ISSUE TITLE
Balnk wallpaper

### ISSUE THREAD

**Issue by mitchjacksontech** [FalconLee1011/Customized-Aerial-Screen-Saver#3]
The desktop wallpaper often goes blank when using a custom arial set by this utility. It also is immediately blank when using the preview function. I have a theory about this. Here's some sippets from `entries.json`

One of the original buit-in video definitions

```json
    {
      "localizedNameKey": "Y011_C001_0305_NAME",
      "showInTopLevel": true,
      "previewImage": "https:\/\/sylvan.apple.com\/itunes-assets\/Aerials126\/v4\/51\/ff\/08\/51ff0824-8da5-78f0-e218-9e61264965bb\/Landscapes_Yosemite_01@2x.png",
      "id": "4109D42A-D717-46A7-A9A2-FE53A82B25C0",
      "shotID": "Y011_C001_0305",
      "accessibilityLabel": "Yosemite",
      "pointsOfInterest": {
        "44": "Y011_C001_44",
        "120": "Y011_C001_120",
        "87": "Y011_C001_87",
        "585": "Y011_C001_585",
        "0": "Y011_C001_0",
        "390": "Y011_C001_390",
        "470": "Y011_C001_470"
      },
      "categories": [
        "A33A55D9-EDEA-4596-A850-6C10B54FBBB5"
      ],
      "preferredOrder": 50,
      "url-4K-SDR-240FPS": "https:\/\/sylvan.apple.com\/itunes-assets\/Aerials116\/v4\/cb\/5b\/50\/cb5b5035-6701-619f-9065-3d7d0e5fbef4\/Y011_C001_UHD_SDR_FRC240fps_sdr_4k_qp22_240p_t2160_tsa.mov",
      "subcategories": [
        "E4B726A1-7DD5-4E5A-A430-C99FE47E7A4F"
      ],
      "includeInShuffle": true
    },
```

Here's an custom item added by this utility

```json
    {
      "shotID": "",
      "localizedNameKey": "Puget Sound",
      "pointsOfInterest": {},
      "showInTopLevel": true,
      "id": "3B6A19F0-DA52-4915-A7B8-DC47D37DF9E3",
      "accessibilityLabel": "Puget Sound",
      "subcategories": [
        "E2B12963-9698-43BE-890D-DDC900C95D00"
      ],
      "categories": [
        "C8EE7C2A-8025-4094-91C2-9637B6D2A64D"
      ],
      "url-4K-SDR-240FPS": "\/Users\/mjac\/Library\/Application Support\/com.xtl.customizedarealscreensaver\/video\/3B6A19F0-DA52-4915-A7B8-DC47D37DF9E3.mov",
      "preferredOrder": 0,
      "previewImage": "\/Users\/mjac\/Library\/Application Support\/com.xtl.customizedarealscreensaver\/asset-perview\/asset-preview-3B6A19F0-DA52-4915-A7B8-DC47D37DF9E3.jpg",
      "includeInShuffle": true
    }
```

This utility does not add a `pointsOfInterest` key. Is it possible when wallpaper gets set by the OS without actually playing the vdieo for a while and pausing on a frame, that it chooses a wallpaper from from the video using this `pointsOfInterest` array? Could these values be pointing at keyframes within the video for the OS to use as wallpaper images when setting wallpaper without actually playing through the video sequence? Maybe that's why the wallpaper keeps going blank? It looks like `shotID` may need a similar string identifying a frame from the video?

**Comment by FalconLee1011** [comment]
@mitchjacksontech The theory you've provided seems possible, I will conduct some experiments based on this theory soon! thanks :))

Also, video itself might be a factor too, the recommended video format is:
| Attribute | Requirements |
|:-|:-| 
| Length | At least 15 seconds |
| Audio Track | Prefer no audio track |
| FPS (Frames Per Second) | At 60 FPS |
| Aspect Ratio | 16:9 |
| Resolution | At least 1920x1080 |
| Video Format | Must be .mov |

**Comment by ian-h-chamberlain** [comment]
@FalconLee1011 that's a handy list of requirements, do you think it makes sense to add that to the README (or if possible, have the app check some of those things when user adds a video)?

I realize a lot of this is just reverse engineering Apple's implementation, so it makes sense that it's not straightforward to figure out what exactly is possible or not. But if you have seen experimentally that some of these things don't work then it would be great to document it!

**Comment by jdrupal-dev** [comment]
Any findings on this?
I have the same issue, where the wallpaper goes black when logging in.

**Comment by plambe** [comment]
I tried adding pointsOfInterest, but it didn't help at all. They seem to be just offsets in seconds from the beginning of the video. 

@FalconLee1011, where is the recommended video format documented? 

I think the format might be the problem.

**Comment by joelklint** [comment]
Hi! 👋 I'm experiencing this issue aswell. Video works upon first screen lock. Upon first unlock everything goes black and stays black. Also the preview in System Preferences is broken from the very start.

I tried transcoding the video to 1920x1080 60 FPS and it still failed. The table below is describes the videos used. 
| Attribute | Requirements |
|--------|--------|
| Length | 19 seconds |
| Audio Track | Silence (not sure if video contains silent audio track or just no track at all) |
| FPS (Frames Per Second) | 60 FPS |
| Aspect Ratio | 16:9 |
| Resolution | 3840 × 2160 and 1920x1080 | 
| Video Format | .mov |

**Comment by YusongLeng** [comment]
Hi all,

I've been experiencing the 'blank wallpaper' issue as well, and have done some experiments to try and find a solution. Here's what I've discovered:

I started by downloading two official video wallpapers from System Settings:
1. **Landscape**/**California’s Temblor Range**: `F390FE3B-FA61-483D-BADC-2447F89951BA.mov`
2. **Landscape**/**Redwoods from Above**: `AA5E82B9-289A-480C-A14B-242989107275.mov`

I then swapped the filenames of these videos without altering their associated files or metadata (e.g. snapshots, `entries.json`). Interestingly, selecting '**California’s Temblor Range**' in System Settings now plays the '**Redwoods from Above**' video, and this wallpaper worked perfectly fine – the video smoothly stopped to a still wallpaper when I unlocked it back from Screensaver.

This led me to hypothesize that the 'blank wallpaper' issue lies within the video files themselves, rather than other system files.

---

With this assumption, my next step is to replicate the video specifications of the official wallpapers by encoding a new video. I first used `ffprobe` to extract the metadata of both official wallpapers, attached below:

1. **Landscape**/**California’s Temblor Range**:
    ```bash
    Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/Library/Application Support/com.apple.idleassetsd/Customer/4KSDR240FPS/F390FE3B-FA61-483D-BADC-2447F89951BA.mov':
      Metadata:
        major_brand     : qt  
        minor_version   : 0
        compatible_brands: qt  
        creation_time   : 2023-08-18T03:36:16.000000Z
      Duration: 00:08:30.51, start: 0.000000, bitrate: 11946 kb/s
      Stream #0:0[0x1](und): Video: hevc (Main 10) (hvc1 / 0x31637668), yuv420p10le(tv, bt709), 3840x2160 [SAR 1:1 DAR 16:9], 11921 kb/s, 239.76 fps, 239.76 tbr, 239760 tbn (default)
        Metadata:
          creation_time   : 2023-08-18T03:36:16.000000Z
          handler_name    : Core Media Video
          vendor_id       : [0][0][0][0]
          encoder         : HEVC
    ```
    
2. **Landscape**/**Redwoods from Above**:
    ```bash
    Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/Library/Application Support/com.apple.idleassetsd/Customer/4KSDR240FPS/AA5E82B9-289A-480C-A14B-242989107275.mov':
      Metadata:
        major_brand     : qt  
        minor_version   : 0
        compatible_brands: qt  
        creation_time   : 2023-08-14T17:52:43.000000Z
      Duration: 00:05:07.24, start: 0.000000, bitrate: 12593 kb/s
      Stream #0:0[0x1](und): Video: hevc (Main 10) (hvc1 / 0x31637668), yuv420p10le(tv, bt709), 3840x2160 [SAR 1:1 DAR 16:9], 12568 kb/s, 239.76 fps, 239.76 tbr, 239760 tbn (default)
        Metadata:
          creation_time   : 2023-08-14T17:52:43.000000Z
          handler_name    : Core Media Video
          vendor_id       : [0][0][0][0]
          encoder         : HEVC
    ```

Using this information, I created a [dummy video](https://drive.google.com/file/d/1_1029E_2T5MRfXJTfISl9Weesko03iPo/view?usp=sharing) to match these specs using `ffmpeg`. Here's the metadata of this dummy video:

```bash
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'clips/slice_01.mov':
  Metadata:
    major_brand     : qt  
    minor_version   : 512
    compatible_brands: qt  
    encoder         : Lavf60.16.100
  Duration: 00:00:15.02, start: 0.000000, bitrate: 37607 kb/s
  Stream #0:0[0x1]: Video: hevc (Main 10) (hvc1 / 0x31637668), yuv420p10le(tv, bt709, progressive), 3840x2160 [SAR 1:1 DAR 16:9], 37583 kb/s, 239.76 fps, 239.76 tbr, 239760 tbn (default)
    Metadata:
      handler_name    : VideoHandler
      vendor_id       : FFMP
      encoder         : Lavc60.31.102 libx265
```

Unfortunately, this dummy video still encountered the 'blank wallpaper' issue, leading me to speculate that discrepancies in `handler_name`, `vendor_id`, `encoder`, or potentially unidentified metadata could be contributing factors.

---

The exact cause of the 'blank wallpaper' issue remains elusive. It suggests a deeper compatibility or encoding problem. Further investigation into metadata differences and additional encoding parameters may be necessary.

I hope these insights prove useful and look forward to any additional information or solutions from the community.

**Comment by chrisly-bear** [comment]
I'm experiencing the same issue. The wallpaper and screensaver work initially, but then go to black when moving from screensaver to desktop and stay that way.

I looked at the output of the Console.app and noticed these two errors:

```log
error	17:22:27.426531+0200	WallpaperVideoExtension	[<private>]: Error handling sample: Error Domain=WallpaperVideoCore.VideoSampleReadingErrors Code=4.
error	17:22:27.426595+0200	WallpaperVideoExtension	[<private>]: VideoPlayerLayer failed to acquire next sample due to error: Error Domain=WallpaperVideoCore.VideoSampleReadingErrors Code=4
```

It sounds like the WallpaperVideoExtension tries to sample a frame from the video (presumably the frame that will be displayed as the static image of the desktop background?). I tried to figure out what the VideoSampleReadingError Code 4 means. The errors seem to be part of a private API, so there's no documentation on it. However, upon disassembling the WallpaperVideoExtension binary (`/System/Library/ExtensionKit/Extensions/WallpaperVideoExtension.appex/Contents/MacOS/WallpaperVideoExtension`) using [Hopper](https://www.hopperapp.com/index.html) I found the following enum:

```objc
enum WallpaperVideoCore.VideoSampleReadingErrors {
  case noAssetReader
  case noSampleBuffer
  case noPixelBuffer
  case noTimeAdjustedSampleBuffer
  case noTemporalInfo
}
```

Assuming these are in the correct order, error 4 would be **noTemporalInfo**, which does sound like there is some metadata missing in the video itself (I didn't find anything interesting in the Aerial.sqlite database, which could've potentially also stored such "temporal information"). However, comparing @YusongLeng's dummy video to one of the Apple videos, I don't see any glaring differences (at least using `ffprobe`, maybe there are unidentified metadata as YusongLeng also suggests?).

I came across this [YouTube video](https://youtu.be/SBFtSX2qddU?si=wnesmNHB-xVHVT9Z&t=99) that mentions the custom video must have the same length as the Apple video it replaces, so I tried replacing the "Tahiti Coast" video with a video of the exact same length (01:24 min). However, this lead to the same VideoSampleReadingError Code 4.

Not sure where to go from here, but hopefully this information is useful to someone looking into the problem (maybe someone with more reverse engineering or video encoding knowledge).

**Comment by karashiiro** [comment]
Also seeing this issue, but as a band-aid for others: I've found I can just use the customized Aerial as a screensaver only (unchecking "Show as wallpaper") and then setting the corresponding still image (the thumbnail, in my case) as my actual wallpaper. Might not look great for more complex Aerials, but it worked for the simple one I used.

**Comment by Sugored** [comment]
Something I discovered:
Videos recorded from iPhones and iPads **will work**, regardless of framerate, resolution, or length. I tested it on videos I recorded from an iPhone 5c, iPhone X, and 2017 iPad Pro and they all slow down and pause properly upon login and logout.

I have a feeling videos exported from Final Cut Pro might work, but I can't test that since I don't have the software.

**Comment by jtpotato** [comment]
> I have a feeling videos exported from Final Cut Pro might work

There's a pretty good reason to believe that iMovie will export in the same format as FCPX (at least... thats what I think, I haven't confirmed it)

I do have a copy of Final Cut, but it will be several weeks before I have the chance to do any thorough investigation

**Comment by Sugored** [comment]
Update:
So I tried it on iMovie and it didn't work, video immediately went black upon login.
I also got a trial of FCP and tried exporting in multiple options that export to MOV, and that also didn't work. Though, since it was the free version I didn't have access to Compressor, which might have more options.

**Comment by vikegart** [comment]
Same here, tried to export video from Final Cut Pro and add some fps to 240 via FrameGUI, but faced with stopping video & black walls

**Comment by Taewan-P** [comment]
> Videos recorded from iPhones and iPads **will work**, regardless of framerate, resolution, or length.

Can this be emulated via an Emulator? Just had a thought that the wallpaper doesn't need sounds, so why not emulate the camera in iPhone or iPad emulator and record it? Any way of doing that?

After searching the internet, I found out that XCode simulators don't support cameras...

**Comment by KiwiJanus** [comment]
This is the same issue as #12 I believe.
Current theory is that it's tied to the wallpaper videos being slow-motion. It's certainly some special type of slow motion, because Quicktime Player adds a secondary slider bar to adjust the speed fade-in and fade-out times.

**Comment by Taewan-P** [comment]
Quicktime slide bar is added when the video is in high frame rate. I tried using frame interpolation techniques to make a 4K240fps video, and in Quicktime it showed the secondary bar. However it was not solved. Still have issues.

**Comment by Proton0** [comment]
> Attribute 	Requirements
> Length 	19 seconds
> Audio Track 	Silence (not sure if video contains silent audio track or just no track at all)
> FPS (Frames Per Second) 	60 FPS
> Aspect Ratio 	16:9
> Resolution 	3840 × 2160 and 1920x1080
> Video Format 	.mov

I found out that videos that are 25FPS work best, they flicker but after a second it shows up again and i dont have an error about missing temporal information in Console.app

**Comment by Kcraft059** [comment]
It actually seems that video from Final Cut Camera - Shot on iPhone do work perfectly. After trying various things (as changing the video codec to match HEVC, and all of the posible things I could do to mimic an original video format) it really seems like that there's some data stored in apple generated .mov (as said in the issue #12 here -> https://github.com/FalconLee1011/Customized-Aerial-Screen-Saver/issues/12#issuecomment-2311595169 ). It is surely related to the apple QuickTime video format, and there seems to be some documentation, but sadly I don't have any good understanding of video codec or anything related to it to exploit it... 
What I found so far : https://developer.apple.com/documentation/quicktime-file-format
And what could __**maybe**__ linked to noTemporalInfo error: https://developer.apple.com/documentation/quicktime-file-format/overview_of_timed_metadata

**Comment by Kcraft059** [comment]
Made research and I got some (maybe wrong) conclusions, the problem is directly due to the atom structure of the file (Specifically in the hvc1 codec - codec string hvc1.2.4), working files possess an atom under moov/mdia/minf/stbl boxes named csgm which isn't documented at all https://mp4ra.org/registered-types/boxes and which contains some random value... (for sure not random but not documented) 

csgm value example for stock apple Sonoma aerial wallpaper :
```
<UnknownBox Size="53" Type="csgm" Specification="unknown" Container="unknown" data="0x000000007473636C00000000000000040000000200000001000000010000001000015180112345545534554550" >
</UnknownBox>

or - in hex 

00000035 6373676D 00000000 7473636C 00000000 00000004 00000002 00000001 00000001 00000010 00015180 11234554 55345545 50
```

when removing it from the file with some dark magic of hex editing you end up with a perfectly working mov file, but which recreate the exact same issue custom non-apple-generated videos have with blank screen and aerial freezing upon relogging. So I think it might be our issue, but I haven't found yet any way to add this atom, with cooherent value to a custom non-apple-generated video, because if you just overwrite the value of the atom with random data it corrupts the file and it's not even playable (which maybe strange but is opposed to removing the atom itself which doesn't affect the playability of the file) - I've linked the edited stock macOS Sonoma wallpaper mov with the csgm atom removed (one of the 2 in fact but one is enough to demonstrate the correlation) https://drive.google.com/file/d/17AJvl8VnQ-QIQR-OJMMNNiOxdrYFDsgk/view?usp=sharing - The wallpaper should run fine once and then upon login again it will stop to be an aerial wallpaper and won't be aerial anymore and just a fixed image.

**Comment by 0xC0FFEEEE** [comment]
The data attributes value from the csgm object - `000000007473636C00000000000000040000000200000001000000010000001000015180112345545534554550` decodes to the following, which could be timescale data:

```
tsclQ#ETU4UEP
```

**Comment by Kcraft059** [comment]
Yes we might have found the missing puzzle piece of this problem, I think we're not far from solving it but, as for now since we can't interpret it it might not be super useful, but we can still hope for someone to find how to create these kind of atoms...
(I also forgot to say that there were 2 csgm atoms in the file the two were  under moov/mdia/minf/stbl)

`5csgmtsclQ#ETU4UEP5csgmtsasQ`

**Comment by KiwiJanus** [comment]
I think we are unlucky and that there are multiple atoms required for them to work.

I've been using mp4dump from Bento4 suite to inspect the atom structure of videos. I noticed that all of the stock wallpapers I checked contained the same basic structure for moov/trak/mdia/minf/stbl/sgpd (I don't know whether the raw hex is the same though)

All videos contained this section:
```
[sgpd] size=12+112, version=1
  grouping_type = tscl
  default_length = 20
  entry_count = 5
  entries:
    (       0) [00 22 fc 00 00 00 f0 00 00 00 00 00 b7 00 00 00 00 00 00 00]
    (       1) [01 22 fc 00 00 00 f0 00 00 00 00 00 b7 00 00 00 00 00 00 00]
    (       2) [02 22 fc 00 00 00 f0 00 00 00 00 00 b7 00 00 00 00 00 00 00]
    (       3) [03 22 fc 00 00 00 f0 00 00 00 00 00 b7 00 00 00 00 00 00 00]
    (       4) [04 22 fc 00 00 00 f0 00 00 00 00 00 b7 00 00 00 00 00 00 00]
```
When I removed this section, and only this section, the stock wallpapers ceased to work, despite being a perfectly fine mov. They would also return a black screen on login, and log the error to the Console app.
Removing even one of the cgsm atoms caused it to fail for me as well.

mp4edit lets you insert and replace atoms, but I can't get it to let me create more than one sgpd or cgsm atoms (stock wallpapers have two of each)

**Comment by Kcraft059** [comment]
Ok as you said I compared (with `mp4box -diso`) the atom structure of sgpd (SampleGrouPDescriptionbox) between 3 stock MacOS Aerial Wallpapers
and it seems that they all contains the same value of 

`<SampleGroupDescriptionBox Size="124" Type="sgpd" Version="1" Flags="0" Specification="p12" Container="stbl traf" grouping_type="tscl" default_length="20">
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,0022FC000000F00000000000B700000000000000"/>
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,0122FC000000F00000000000B700000000000000"/>
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,0222FC000000F00000000000B700000000000000"/>
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,0322FC000000F00000000000B700000000000000"/>
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,0422FC000000F00000000000B700000000000000"/>
</SampleGroupDescriptionBox>

<SampleGroupDescriptionBox Size="28" Type="sgpd" Version="1" Flags="0" Specification="p12" Container="stbl traf" grouping_type="tsas" default_length="0">
<DefaultSampleGroupDescriptionEntry size="0" data="data:application/octet-string,"/>

or in hex respectively for the first sgpd entry

|sgpdtscl"üð·"üð·"üð·"üð·"üð·

0000007C 73677064 01000000 7473636C 00000014 00000005 0022FC00 0000F000 00000000 B7000000 00000000 0122FC00 0000F000 00000000 B7000000 00000000 0222FC00 0000F000 00000000 B7000000 00000000 0322FC00 0000F000 00000000 B7000000 00000000 0422FC00 0000F000 00000000 B7000000 00000000

and for the second 

sgpdtsas

0000001C 73677064 01000000 74736173 00000000 00000001 00000000

`

And I repeat it these values does not change and can be implemented manually via hex editing. (you will ofc need to change all size of parent boxes, but in theory it could be absolutely possible). But keep in mind that these values are the same for mov files with HEVC codec specifically at 239.70 fps and that these value could be tied to it, so if tested I recommend to use the hvcC codec and the same frame rate to be sure it doesn't mess it up 

Ps: I checked the value for a working (as aerial wallpaper it works perfectly fine with any issue of blank screen nor freeze) file created with my iPhone using Final Cut camera at 30 fps (1920x1080) and encoded with hvc1 (H.265) and it has different value for this parameter:

`</SampleDescriptionBox>
<SampleGroupDescriptionBox Size="104" Type="sgpd" Version="1" Flags="0" Specification="p12" Container="stbl traf" grouping_type="tscl" default_length="20">
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,000200000000B380000000007B00000000000000"/>
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,010200000000B380000000007B00000000000000"/>
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,000220000000B000000000007B00000000000000"/>
<DefaultSampleGroupDescriptionEntry size="20" data="data:application/octet-string,010220000000B000000000007B00000000000000"/>
</SampleGroupDescriptionBox>
<SampleGroupDescriptionBox Size="26" Type="sgpd" Version="1" Flags="0" Specification="p12" Container="stbl traf" grouping_type="sync" default_length="1">
<SyncSampleGroupEntry NAL_unit_type="20"/>
<SyncSampleGroupEntry NAL_unit_type="21"/>
</SampleGroupDescriptionBox>
<SampleGroupDescriptionBox Size="28" Type="sgpd" Version="1" Flags="0" Specification="p12" Container="stbl traf" grouping_type="tsas" default_length="0">
<DefaultSampleGroupDescriptionEntry size="0" data="data:application/octet-string,"/>
</SampleGroupDescriptionBox>

or in hex respectively for the first entry 

hsgpdtscl³{³{ °{ °{

00000068 73677064 01000000 7473636C 00000014 00000004 00020000 0000B380 00000000 7B000000 00000000 01020000 0000B380 00000000 7B000000 00000000 00022000 0000B000 00000000 7B000000 00000000 01022000 0000B000 00000000 7B000000 00000000

and for the second 

sgpdsync

0000001A 73677064 01000000 73796E63 00000001 00000002 1415`

**Comment by KiwiJanus** [comment]
Given that the errors potentially point towards temporal info, I has perfectly matched the framerate, timescale in mdhd atom, and even modified and compiled ffmpeg so that I could get the mvhd atom timescale to be 600. I has used this for the closest approximation to the stock wallpaper I could get:
```
ffmpeg -i "input.mp4" -map_metadata -1 -c:v libx265 -preset medium -pix_fmt yuv420p10le -x265-params "level-idc=6.1:no-info=1" -tag:v hvc1 -filter:v fps=fps=239.760 -video_track_timescale 239760 "timescale.mov"
```
Having now also done testing with Final Cut Camera, I see that I could have saved myself that effort.

Slow motion videos taken straight from an iPhone 15 Pro work to some extent. 
Ones taken from the stock camera app gave me a number of additional problems, but I too have found that those filmed with Final Cut Camera work. 30fps did not work for me, but filming in either 120fps or 240fps did work (SDR HEVC).

I guess that means we probably need to correctly set the two sgpd and two csgm atoms at the very least. This is beyond my ability :(.
- sgpd[0] appears to be consistent between videos of the same type. It was the same between various 4k 240fps aerials. It was also the same between two 1080p 240fps Final Cut videos (but different to that of the aerials). It also changed between the 240fps and 120fps final cut videos, so @Kcraft059 is probably right in that it could very well be tied to frame rate and such.
- sgpd[1] appeared to be the same between all of them.
- Both csgm atoms were unique for every file.

I too am now lost on how to proceed.

**Comment by 0xC0FFEEEE** [comment]
It looks like it may be possible to add custom atoms using `mp4box`.

I don't have much time to dig in to this today but the patch command test suite can be found [here](https://github.com/gpac/testsuite/blob/261434761a44b277159b223264cf1b2fc451b084/scripts/mp4box-boxpatch.sh#L9) and example patch xml files [here](https://github.com/gpac/testsuite/tree/261434761a44b277159b223264cf1b2fc451b084/media/boxpatch).

Sadly the mp4box dump command (`mp4box -diso <file>`) complains about the csgm atoms and dumps seemingly empty objects for those two atoms in the resulting XML file.

**Comment by martinn000** [comment]
This issue is still present on MacOS Tahoe using the new beta version :()

**Comment by Kcraft059** [comment]
Yeah, saddly I think it's on purpose, the video files used by apple for their aerials aren't quite "standard" video files, it's not an apple problem, their software works perfectly fine it's just that it's not intended to be used as we do right now. And ofc apple fashion way of doing it, is to not document it at all because, why bother when your goal is to restrain user from doing whatever they want with their devices… anyway thx Apple <3.

**Comment by FalconLee1011** [comment]
Hi everyone — again, thank you all for using and supporting this project!

As @Kcraft059 pointed out, it’s common for Apple to design things to work seamlessly within their own ecosystem. While this approach helps keep the platform secure and stable, it can create roadblocks for us when trying to “hack around” or customize it beyond their intended use.

That said, I’ll be introducing a new “Encode” feature in the upcoming v2.0 release. This will help you convert videos into an Apple-friendly format that’s much more likely to work reliably with the system.

Thanks again for your patience and feedback!

**Comment by Kcraft059** [comment]
> That said, I’ll be introducing a new “Encode” feature in the upcoming v2.0 release. This will help you convert videos into an Apple-friendly format that’s much more likely to work reliably with the system.


Are you kidding me ? No but like seriously... that would be actually game breaking... respect, and thx for your dedication ! (Just a bit interested on actually how you'll do to make it work) Because having worked a bit the subject by myself it seems really hard to reproduce... 
But anyway: Good luck ! 

_(Will it be in the public beta yet or to be implemented ?)_

**Comment by Kcraft059** [comment]
Update : 

This app seems to have found a way to make a video work as a Macos screensaver : http://cindori.com/backdrop (by using assets aerials) - idk how they did, but it seems it works for them, could be an idea to investigate.

**Comment by KiwiJanus** [comment]
Yeah, the dev for Backdrop said they've got a custom video encoding process on their [https://www.reddit.com/r/macapps/s/plIHS7XejS](Reddit post) so they've probably managed to figure out how to add those pesky undocumented atoms needed to make the video play smoothly.

**Comment by OllieFallon2** [comment]
I ended up using Backdrop like others here, and its worked perfectly, can either use it by itself or create a custom wallpaper, set it then copy the MOV from idleassetsd into Customized Aerial Screen Saver and it works perfectly, pauses and resumes without crashing. Thanks to all for the suggestion of Backdrop! Unfortunately does cost money but I'd argue its worth it.
